### ISSUE TITLE
Simultaneous creation on call to findOneOrCreate() causes fatal error?

### ISSUE THREAD

**Issue by benguild** [propelorm/Propel2#1180]
I've found that if another process creates a row between the SELECT and INSERT calls of `findOneOrCreate()`, the script crashes for "Uncaught exception 'PDOException' with message 'SQLSTATE[23000]: Integrity constraint violation: 1062 Duplicate entry". This doesn't really make sense in a `findOneOrCreate()` situation, no?

To me, it seems like both an edge case and undesirable behavior for the crash to occur. Yet, wrapping absolutely every single `findOneOrCreate()` with a try/catch seems stupid... yet also a valid workaround. At least for the ones with possible collisions.

Shouldn't it do SELECT -> INSERT -> SELECT ... **then** throw any exception from the INSERT if the row still does not exist?

**Comment by marcj** [comment]
You get "1062 Duplicate entry", so why should it throw a exception stating that "the row still does not exist?"? In fact, "duplicate entry" means this row already exists.

**Comment by benguild** [comment]
So, why wouldn't it re-SELECT to find the row?

**Comment by marcj** [comment]
Because it's not the job of an ORM to sustain parallelism. Obviously you have so high traffic, that chances are high that race conditions happens, so you need to make sure (through table locking or different keys perhaps) that such stuff can not happen. Propel should really not re-select when a insert fails due to duplicate key, because duplicate key means only that the key is already there - it is not a check whether the actual wanted object value is currently in the database (with the column values given by the filterBy in the query object). Checking this and reload would result in undesirable behavior. So, I highly recommend to build an architecture that can handle multiple client requests simultaneously - it's unfortunately not the job of an ORM.

So when you have code in your application that has a high chance running parallel you should add (table-) locking instead of fixing symptoms with try-catch/re-tries. That's the only way to make sure that two processes don't handle the same database row.

**Comment by benguild** [comment]
Fair enough, I guess that's a valid point and consideration. I definitely agree that not every case needs the SELECT/INSERT/SELECT loop, but in the case of a 1062 it does create some confusion in such an abstracted setting.

**Comment by benguild** [comment]
Reopening because I’m actually running into this issue all over this prototype application.

The issue isn’t just “high-traffic only” situations, but more in settings where there may be a lot of asynchronous operations. In this case, LOCK’ing a table constantly for every single `findOneOrCreate()` does not seem to make sense, given that a lot of rows may be INSERT'ed into said table at once from separate operations and it's possible to have a key collision from another operation that succeeded before the current one. Regardless of anyone's opinion, that's entirely possible in a normal setting for asynchronous applications.

Would it be appropriate to add, for example, a method like this:
```
$object->insertWhereNot($query);
```

... Which might return a BOOL on success/failure, and update the internal primary key pointer and isNew() status based on success? The SQL generated/executed (instead of a plain INSERT which would fail) would be something like:
```
INSERT INTO `table` (`columnA`, `columnB`) SELECT "abc" AS `columnA`, "abc" as `columnB` FROM DUAL WHERE NOT EXISTS (FIND QUERY)
```

This way, you could basically change your workflow to be like:
```
$object=null;

for ($i=0; $i<2; $i++)
{
    $object=$query->find(($i ? $writeConnection : $readConnection)); // On second attempt, use write connection as it was most likely INSERT'ed very recently by another process.
    
    if (!$i && !$object)
    {
        $newObject=(new Object());

        if ($newObject->insertWhereNot($query))
        {
            $object=$newObject;
            break;
            
        }
        
    }

}

if (!$object)
{
    // Throw exception.
    
}
```

... This routine could also be bundled into another function, for example, to abstract this issue away entirely:

```
$object->asynchronousSaveIfNotExists();
```

... Or, something?

**Comment by marcj** [comment]
If you really can't make sure beforehand that you insert a valid object with a unique key you should try using [upserts](http://dev.mysql.com/doc/refman/5.7/en/insert-on-duplicate.html).

**Comment by benguild** [comment]
1. Does Propel support those? If so, wouldn't they be embedded in `findOneOrCreate()`
2. In my experience, those tend to bump the AUTO_INCREMENT value by 1 even on failure, which can result in runaway indices.

Also, I don't understand why the above method is not a good one. How else can you make sure data exists in the table?

**Comment by marcj** [comment]
Both methods are not implemented in Propel.

> Also, I don't understand why the above method is not a good one.

It's a elegant one. Whether it's good is based on other stuff: a) Does it work cross-db (sqlite, postgres, mysql), b) is it easy to implement, c) do people understand what this is doing.

> How else can you make sure data exists in the table?

Locking. This is what locking is made for. Since most databses uses mutex locking is usually very fast.

I know, a atomic operation like upsert or your proposal is even faster, however I like more the obvious way. Whatever the best way is: I suggested the locking way because this is already possible with Propel. All other approaches need first a adjustment in Propel's core, which I really can not support at the moment. If you want to give it a try I'll support you, but I can't hack upsets or insert-as-select in Propel right now. :)

**Comment by benguild** [comment]
I understand, that makes sense. Do you have a link to a demo to how you'd recommend using Propel's locking methods around a `findOneOrCreate()`? I'm assuming you'd use a WRITE only lock, of course.

============================

For the INSERT...WHERE NOT EXISTS discussion, the trickiness with this actually comes from different versions of MySQL. I know that Postgres and SQLite support the syntax, but it's slightly different in MSSQL I believe. (e.g. "IF NOT (SELECT) INSERT..." or something, I'm not sure.)

The "DUAL" keyword was a recent addition in newer MySQL versions, I think, and detecting this could be troublesome. On older versions, DUAL was not expected as far as I know. Therefore, on newer versions, it'll fail, but older ones it would not. This, most likely, would require a configuration flag in the Propel configuration and could prove too troublesome for some users.

**Comment by benguild** [comment]
Also, I'm happy to prototype out this function for my own use and patch it into Propel. I haven't really looked into that yet, pretty tired.

**Comment by marcj** [comment]
> but it's slightly different in MSSQL I believe

I don't give a ...;) If not easily supported by MSSQL and Oracle, then with those adapter it's just not possible. Ignoring mssql/oracle is easier as the userbase for those adapter is extremely low.

> The "DUAL" keyword was a recent addition in newer MySQL versions 

Isn't it possible to use the table name instead of `dual`?


```php
$con = Propel::getWriteConnection();
$con->exec('LOCK TABLES foo WRITE');
$con->beginTransaction(); //necessary because propel already uses transaction in ->save(), which would unlock all locks implicity
try {
   //do some stuff
   $con->commit();
catch (\Exception $e) {
   $con->rollback();
   throw $e;
}
$con->exec('UNLOCK TABLES');
```

Please note (since you mentioned async php): locks are session based, not transaction based. So you can't make it work when your have async operations that use the same mysql connection.

**Comment by benguild** [comment]
So, is there no way to use table write locking without dropping down to raw SQL? In that case I might as well call raw SQL for the INSERT.

**Comment by marcj** [comment]
Locking is not part of the ORM API, so you need to execute custom queries, yes. If you want to build the insert query on your own, feel free. I guess the locking is still more readable.

**Comment by benguild** [comment]
But, I take it you could wrap the raw LOCK query around a typical new()/save() procedure as long as it's on the same connection?

=================================

Also, I don't understand what you mean here:
> Please note (since you mentioned async php): locks are session based, not transaction based. So you can't make it work when your have async operations that use the same mysql connection.

The best example I can think of is that the user, for example, presses the Refresh button in their browser twice or double-clicks the Submit button by accident. We have to "de-bounce" this behavior, rather than return a fatal error to the client when the data was INSERT'ed correctly the first time.

Ideally, wrapping the SELECT/INSERT combo in a WRITE LOCK transaction for each request would accomplish:
1. The AUTO_INCREMENT value would not increase for the second attempt.
2. The first would SELECT then INSERT, where as the second would SELECT only.

I'm not sure about any sync lag with the read and write DB boxes, which is why I added the fallback loop in my code above (to check the write box before INSERT'ing after a SELECT fail or whatever), but maybe that's a non-issue.

================================

Also, if I do write my own INSERT...WHERE NOT EXISTS (QUERY) approach in raw SQL, what's the best way for me to turn a Propel query object into SQL ready query that just needs bindable parameters? Is that supported...? A way to get the generated query and an array of parameters/keys?

**Comment by marcj** [comment]
```php
$query = new FooQuery();
$params = [];
$query->addSelfSelectColumns();
$query->configureSelectColumns();
$query->createSelectSql($params);
```

**Comment by benguild** [comment]
So how do I get the parameters out if I do:```$query->filterByColumnABC("some value")```






On Wed, May 4, 2016 at 4:07 PM -0700, "Marc J. Schmidt" <notifications@github.com> wrote:










$query = new FooQuery();
$params = [];
$query->addSelfSelectColumns();
$query->configureSelectColumns();
$query->createSelectSql($params);

**Comment by benguild** [comment]
So how do I get the parameters out if I do something like this preceding the `createSelectSql()` call:

```
$query->filterByColumnABC("some value")
```

**Comment by benguild** [comment]
Is this valid? Instead of...
```
$query->filterByColumnABC("some value");
```

Do:
```
$query->filterByColumnABC("?");
$query->createSelectSql(array("some value"));
```

...? Not sure how else to build queries using the methods.

**Comment by marcj** [comment]
@benguild, no not valid.

**Comment by dereuromark** [comment]
Is this still relevant or can it be closed?
